### PR TITLE
Retain external transitions in envelopes

### PR DESCRIPTION
**Pull request by deepthiskumar**
Retain external transitions in envelopes until they are added to the transition frontier. This is required because validations are done in different stages

Checklist:

- [ ] Tests were added for the new behavior
- [X] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

### CODE REVIEW

File: src/lib/envelope/envelope.ml [hunk 1]
```diff
@@ -11,6 +11,13 @@ module Sender = struct
       end
 
       include T
+
+      let equal t t' =
+        match (t, t') with
+        | Local, Local -> true
+        | Remote p, Remote p' -> Peer.Stable.V1.compare p p' = 0
```
Comment (psteckler) [inline src/lib/envelope/envelope.ml]: Can we add `Peer.equal` and use that?

File: src/lib/envelope/envelope.ml [hunk 2]
```diff
@@ -40,6 +47,9 @@ module Incoming = struct
       end
 
       include T
+
+      let equal ~equal_data t t' =
```
Comment (psteckler) [inline src/lib/envelope/envelope.ml]: Should this have a different name, since in general, we have `equal : t -> t -> bool` as a signature.

**Comment by johnwu93** [comment]
Can we have 
```
( ( external_transition_verified Envelope.Incoming.t
                            , state_hash )
                            With_hash.t
                          , state_hash )
```

rather than,

```
( ( external_transition_verified, state_hash )
                            With_hash.t) Envelope.Incoming.t
```

It seems more natural to me. Because you want to everything around a envelope. If you want the most outer layer to be `With_hash`, then it could convey that the hash include `peer` ip address.

**Comment by cmr** [comment]
The hash doesn't include the peer address.

**Comment by cmr** [comment]
@johnwu93 you mean the other way around? the way you request seems to be the way it already is. I agree that we should put the Envelope around the With_hash, not wrap the envelope with a With_hash.

File: src/lib/transition_handler/catchup_scheduler.ml [hunk 3]
```diff
@@ -48,7 +52,9 @@ module Make (Inputs : Inputs.S) = struct
               timeouts. *)
     ; parent_root_timeouts: unit Time.Timeout.t State_hash.Table.t
     ; breadcrumb_builder_supervisor:
-        ( (External_transition.Verified.t, State_hash.t) With_hash.t
+        ( ( External_transition.Verified.t Envelope.Incoming.t
```
Comment (johnwu93) [inline src/lib/transition_handler/catchup_scheduler.ml]: this type is too big and verbose. Can we have a type alias for it. Like `type enveloped_transition = ...`

File: src/lib/transition_handler/catchup_scheduler.ml [hunk 4]
```diff
@@ -109,7 +121,8 @@ module Make (Inputs : Inputs.S) = struct
 
   let has_timeout t transition =
     Hashtbl.mem t.parent_root_timeouts
-      ( With_hash.data transition |> External_transition.Verified.protocol_state
+      ( With_hash.data transition |> Envelope.Incoming.data
+      |> External_transition.Verified.protocol_state
```
Comment (johnwu93) [inline src/lib/transition_handler/catchup_scheduler.ml]: Use External_transition.Verified.parent_hash

File: src/lib/transition_handler/validator.ml [hunk 5]
```diff
@@ -110,6 +113,7 @@ module Make (Inputs : Inputs.With_unprocessed_transition_cache.S) :
                  Perf_histograms.add_span ~name:hist_name
                    (Core_kernel.Time.diff (Core_kernel.Time.now ())
                       transition_time) ;
+                 (*Deepthi: This needs to be in an envelope*)
```
Comment (johnwu93) [inline src/lib/transition_handler/validator.ml]: Delete comments

File: src/lib/transition_handler/processor.ml [hunk 6]
```diff
@@ -38,14 +38,18 @@ module Make (Inputs : Inputs.With_unprocessed_transition_cache.S) :
 
   let run ~logger ~time_controller ~frontier
       ~(primary_transition_reader :
-         ( (External_transition.Verified.t, State_hash.t) With_hash.t
+         ( ( External_transition.Verified.t Envelope.Incoming.t
```
[the remark this answers is not in the document]
Comment (deepthiskumar) [reply src/lib/transition_handler/processor.ml]: Are you referring to the type of `Cache.t`?

File: src/lib/envelope/envelope.ml (hunk 2 above)
Comment (deepthiskumar) [reply src/lib/envelope/envelope.ml]: I was getting an error which I misunderstood. I just had to add the dependency in the dune file. Will use the ppx_derive

File: src/lib/transition_handler/processor.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnwu93) [reply src/lib/transition_handler/processor.ml]: oh yah, I meant the cache. I thought it was a tuple but `Cache.t` is parameterized by two types. Yah, the type of `Cache.t` can make things very confusing.

File: src/lib/ledger_catchup/ledger_catchup.ml [hunk 7]
```diff
@@ -149,8 +146,8 @@ module Make (Inputs : Inputs.S) :
   let materialize_breadcrumbs ~frontier ~logger ~peer
       (Rose_tree.T (foreign_transition_head, _) as tree) =
     let initial_state_hash =
-      With_hash.data (Cached.peek foreign_transition_head)
-      |> Envelope.Incoming.data |> External_transition.Verified.protocol_state
+      Envelope.Incoming.data (Cached.peek foreign_transition_head)
+      |> With_hash.data |> External_transition.Verified.protocol_state
```
Comment (johnwu93) [inline src/lib/ledger_catchup/ledger_catchup.ml]: Use `External_transition.Verified.parent_hash`.

File: src/lib/ledger_catchup/ledger_catchup.ml [hunk 8]
```diff
@@ -192,10 +189,8 @@ module Make (Inputs : Inputs.S) :
                External_transition.Verified.protocol_state)
       in
       let verified_transition_with_hash_enveloped =
-        { verified_transition_with_hash with
-          data=
-            Envelope.Incoming.wrap ~data:verified_transition_with_hash.data
-              ~sender:(Envelope.Incoming.sender transition_enveloped) }
+        Envelope.Incoming.wrap ~data:verified_transition_with_hash
+          ~sender:(Envelope.Incoming.sender transition_enveloped)
```
Comment (johnwu93) [inline src/lib/ledger_catchup/ledger_catchup.ml]: Would using `Envelope.Incoming.map` be better?

File: src/lib/transition_handler/processor.ml [hunk 9]
```diff
@@ -26,30 +26,25 @@ module Make (Inputs : Inputs.With_unprocessed_transition_cache.S) :
...
-  open Consensus
   module Catchup_scheduler = Catchup_scheduler.Make (Inputs)
 
   (* TODO: calculate a sensible value from postake consensus arguments *)
   let catchup_timeout_duration = Time.Span.of_ms 6000L
 
-  let transition_parent_hash t =
-    External_transition.Verified.protocol_state t
-    |> Protocol_state.previous_state_hash
+  let transition_parent_hash = External_transition.Verified.parent_hash
```
Comment (johnwu93) [inline src/lib/transition_handler/processor.ml]: Can you inline `transition_parent_hash`?

**Pull request merged**